src/drivers/kmdf/libraries/KmdfDrivers.qll

```diff
@@ -2,40 +2,3165 @@
+
+class KmdfEVTWdfUsbReadersFailed extends KmdfCallbackRoutine {
+  KmdfEVTWdfUsbReadersFailed() { callbackType.getName().matches("EVT_WDF_USB_READERS_FAILED") }
+}
+
+class KmdfEVTWdfWmiInstanceQueryInstance extends KmdfCallbackRoutine {
+  KmdfEVTWdfWmiInstanceQueryInstance() {
+    callbackType.getName().matches("EVT_WDF_WMI_INSTANCE_QUERY_INSTANCE")
+  }
+}
+
+class KmdfEVTWdfWmiInstanceSetInstance extends KmdfCallbackRoutine {
+  KmdfEVTWdfWmiInstanceSetInstance() {
+    callbackType.getName().matches("EVT_WDF_WMI_INSTANCE_SET_INSTANCE")
+  }
+}
+
+class KmdfEVTWdfWmiInstanceSetItem extends KmdfCallbackRoutine {
+  KmdfEVTWdfWmiInstanceSetItem() { callbackType.getName().matches("EVT_WDF_WMI_INSTANCE_SET_ITEM") }
+}
+
+class KmdfEVTWdfWmiInstanceExecuteMethod extends KmdfCallbackRoutine {
+  KmdfEVTWdfWmiInstanceExecuteMethod() {
+    callbackType.getName().matches("EVT_WDF_WMI_INSTANCE_EXECUTE_METHOD")
+  }
+}
+
+class KmdfEVTWdfWmiProviderFunctionControl extends KmdfCallbackRoutine {
+  KmdfEVTWdfWmiProviderFunctionControl() {
+    callbackType.getName().matches("EVT_WDF_WMI_PROVIDER_FUNCTION_CONTROL")
+  }
+}
+
+class KmdfEVTWdfWorkitem extends KmdfCallbackRoutine {
+  KmdfEVTWdfWorkitem() { callbackType.getName().matches("EVT_WDF_WORKITEM") }
+}
+
+/**
+ * A KMDF callback routine, defined by having a typedef in its definition
+ * that matches the standard KMDF callbacks.
+ */
+class KmdfCallbackRoutine extends Function {
+  /** The typedef representing what callback this is. */
+  KmdfRoleTypeType callbackType;
+
+  KmdfCallbackRoutine() {
+    exists(FunctionDeclarationEntry fde |
+      fde.getFunction() = this and
+      fde.getTypedefType() = callbackType and
+      fde.getFile().getAnIncludedFile().getBaseName().matches("%wdf.h")
+    )
+  }
+}
+
+/** The DeviceAdd callback.  Its callback typedef is "EVT_WDF_DRIVER_DEVICE_ADD". */
+class KmdfEvtDriverDeviceAdd extends KmdfCallbackRoutine {
+  KmdfEvtDriverDeviceAdd() { callbackType.getName().matches("EVT_WDF_DRIVER_DEVICE_ADD") }
+}
+
+/**
+ * A KMDF Function. This class is incomplete.  It is used to represent a Function in the KMDF library.
+ */
+class KmdfFunc extends Function {
+  KmdfFunc() { this.getFile().getBaseName().matches("wdf%.h") }
+}
+
+/**
+ * A KMDF Struct. This class is incomplete.  It is used to represent a Struct in the KMDF library.
+ */
+class KmdfStruct extends Struct {
+  KmdfStruct() { this.getFile().getBaseName().matches("wdf%.h") }
+}
+
+/* KMDF Functions */
+/**
+ * The KMDF DriverEntry function.  KMDF enforces that the function is named DriverEntry.
+ * Additionally, the driver may use the DRIVER_INIRIALIZE typedef.
+ */
+class KmdfDriverEntry extends KmdfFunc {
+  KmdfDriverEntry() { this.getName().matches("DriverEntry") }
+}
+
+class KmdfRtlAssert extends KmdfFunc {
+  KmdfRtlAssert() { this.getName().matches("RtlAssert") }
+}
+
+class KmdfWdfChildListCreate extends KmdfFunc {
+  KmdfWdfChildListCreate() { this.getName().matches("WdfChildListCreate") }

```

Comment:
So it's one thing to have the individual callbacks, since we know we'll be doing reasoning based on them and may add additional information to them over time, but I don't know that having an individual class for every function in wdf.h is adding much value.  In particular, unless you put them in a module, I can see this making the autocomplete very overstuffed when writing anything that starts with "Kmdf".